pkg/controller/networkpolicy/networkpolicy_controller_perf_test.go

```diff
@@ -333,3 +434,75 @@ func BenchmarkSyncAddressGroup(b *testing.B) {
 		c.syncAddressGroup(key.(string))
 	}
 }
+
+func BenchmarkInitXLargeScaleWithSmallNamespaces(b *testing.B) {
+	namespaces, networkPolicies, pods := getXLargeScaleWithSmallNamespaces()
+	benchmarkInit(b, namespaces, networkPolicies, pods)
+}
+
+func BenchmarkInitXLargeScaleWithLargeNamespaces(b *testing.B) {
+	namespaces, networkPolicies, pods := getXLargeScaleWithLargeNamespaces()
+	benchmarkInit(b, namespaces, networkPolicies, pods)
+}
+
+func BenchmarkInitXLargeScaleWithOneNamespace(b *testing.B) {
+	namespaces, networkPolicies, pods := getXLargeScaleWithOneNamespace()
+	benchmarkInit(b, namespaces, networkPolicies, pods)
+}
+
+func BenchmarkInitXLargeScaleWithNetpolPerPod(b *testing.B) {
+	namespaces, networkPolicies, pods := getXLargeScaleWithNetpolPerPod()
+	benchmarkInit(b, namespaces, networkPolicies, pods)
+}
+
+func BenchmarkInitXLargeScaleWithClusterScopedNetpol(b *testing.B) {
+	namespaces, networkPolicies, pods := getXLargeScaleWithClusterScopedNetpol()
+	benchmarkInit(b, namespaces, networkPolicies, pods)
+}
+
+func benchmarkInit(b *testing.B, namespaces []*corev1.Namespace, networkPolicies []*networkingv1.NetworkPolicy, pods []*corev1.Pod) {
+	disableLogToStderr()
+
+	objs := toRunTimeObjects(namespaces, networkPolicies, pods)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	_, c := newControllerWithoutEventHandler(objs...)
+	c.informerFactory.Start(stopCh)
+	c.informerFactory.WaitForCacheSync(stopCh)
+
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	go c.groupingInterface.Run(stopCh)
+
+	for _, namespace := range namespaces {
+		c.groupingInterface.AddNamespace(namespace)
+	}
+	for _, pod := range pods {
+		c.groupingInterface.AddPod(pod)
+	}
+	for _, networkPolicy := range networkPolicies {
+		c.addNetworkPolicy(networkPolicy)
+	}
+	for c.appliedToGroupQueue.Len() > 0 {
+		key, _ := c.appliedToGroupQueue.Get()
+		c.syncAppliedToGroup(key.(string))
+		c.appliedToGroupQueue.Done(key)
+	}
+	for c.internalNetworkPolicyQueue.Len() > 0 {
+		key, _ := c.internalNetworkPolicyQueue.Get()
+		c.syncInternalNetworkPolicy(key.(string))
+		c.internalNetworkPolicyQueue.Done(key)
+	}
+	for c.addressGroupQueue.Len() > 0 {
+		key, _ := c.addressGroupQueue.Get()
+		c.syncAddressGroup(key.(string))
+		c.addressGroupQueue.Done(key)
+	}
```

Comment:
in practice (actual controller), these would be handled by multiple goroutines right?

Reply:
Yes, making the test execute serial is to measure the time and memory accurately when the events are handled in ideal order.